README.md

```diff
@@ -0,0 +1,400 @@
+{
+    fn coerce_unsized(self) -> Pin<U> {
+        Pin {
+            pointer: self.pointer.coerce_unsized(),
+        }
+    }
+}
+```
+
+We should rather strife to have the following 2 implementations:
+```rs
+// Permit going from Pin<impl Unpin> to Pin<impl Unpin>
+impl<P, U> CoerceUnsized<Pin<U>> for Pin<P>
+where
+    P: CoerceUnsized<U>,
+    P: Deref<Target: Unpin>,
+    U: Deref<Target: Unpin>,
+{
+    fn coerce_unsized(self) -> Pin<U> {
+        Pin::new(self.pointer.coerce_unsized())
+    }
+}
+// Permit going from Pin<impl Pin> to Pin<impl Pin>
+impl<P, U> CoerceUnsized<Pin<U>> for Pin<P>
+where
+    P: CoerceUnsized<U>,
+    P: core::ops::Deref<Target: !Unpin>,
+    U: core::ops::Deref<Target: !Unpin>,
+{
+    fn coerce_unsized(self) -> Pin<U> {
+        // SAFETY: The new unpin Pin is derived from another unpin Pin, so we the pinned contract is kept up
+        unsafe { Pin::new_unchecked(self.pointer.coerce_unsized()) }
+    }
+}
+```
+
+
+While this is a breaking change, it should be in line with being a soundness fix.
+Unfortunately, these kind of impl requires negative bounds and negative reasoning which is its own can of worms and therefor likely not to happen, see https://github.com/rust-lang/rust/issues/42721.
+Though maybe allowing them for auto traits alone could work out fine, given those types of traits are already rather special.
+
+## Custom Reborrows
+
+See the following issue for context: https://github.com/rust-lang/rfcs/issues/1403
+
+In the linked issue the idea was to generalize `CoerceUnsized` as a general `Coerce` trait.
+
+The design proposed here would still enable this, although it raises some questions.
+- For one, reborrows should be guaranteed to be no-ops, as all that should change in "reborrow coercions" are the corresponding lifetimes, yet this RFC exposes a function that will be run on coercion.
+- Generalizing this to a general `Coerce` trait would require specialization and/or negative trait bound reasoning, such that `&'a mut T: CoerceUnsized<&'b T>` (for reborrows) but also `&'a mut T: CoerceUnsized<&'b U>, T: Unsize<U>` for unsizing coercions can both be done as impls.
+
+The first issue is only of concern with the proposed design here, while the second one is a more general issue relevant to impl overlap.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- This proposal allows for some non-sensical `CoerceUnsized` implementations resulting in odd unsizing coercions, though the restrictions on the `CoerceUnsized` trait try to limit them (for the time being).
+- Unsizing coercions are now able to run arbitrary user code, placing it into a similar category to `Deref` in that regard, effectively adding yet more user facing \*magic\* to the language.
+- The `Unsize` trait now depends on the `Pointee` trait which means any push for stabilization will depend on the stabilization of said trait.
```

Comment:
I first read "of sad trait" and was like "oh no, poor trait, why are you sad" xD